plutus-core/plutus-ir/src/PlutusIR/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,198 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns        #-}
+module PlutusIR.Transform.CaseOfCase where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Arity
+import PlutusIR.Contexts
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (cons)
+import PlutusIR.Analysis.Builtins
+import PlutusIR.Analysis.VarInfo
+
+{-|
+Perform the case-of-case transformation. This pushes
+case expressions into the case branches of other case
+expressions, which can often yield optimization opportunities.
+
+Example:
+@
+    case (case s of { C1 a -> x; C2 b -> y; }) of
+      D1 -> w
+      D2 -> z
+
+    -->
+
+    case s of
+      C1 a -> case x of { D1 -> w; D2 -> z; }
+      C2 b -> case y of { D1 -> w; D2 -> z; }
+@
+-}
+caseOfCase ::
+    forall m tyname name uni fun a.
+    ( Ord fun, PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , PLC.MonadQuote m
+    ) =>
+    BuiltinsInfo uni fun ->
+    Bool ->
+    Term tyname name uni fun a ->
+    m (Term tyname name uni fun a)
+-- See Note [Case-of-case and conapps]
+caseOfCase binfo conappsOnly t0 = do
+    -- We are going to record information about variables in a global map, so we
+    -- need global uniqueness
+    t <- PLC.rename t0
+    let vinfo = termVarInfo t
+    pure $ transformOf termSubterms (processTerm binfo vinfo conappsOnly) t
+
+processTerm ::
+    forall tyname name uni fun a .
+    (Ord fun, PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique) =>
+    BuiltinsInfo uni fun ->
+    VarsInfo tyname name ->
+    Bool ->
+    Term tyname name uni fun a ->
+    Term tyname name uni fun a
+processTerm binfo vinfo conappsOnly t
+    -- We have a saturated datatype matcher application
+    | Just (SplitMatchContext outerVars (outerScrut, outerScrutAnn) (outerResTy, outerResTyAnn) outerBranches, _, reconstructOuter) <- splitMatch binfo vinfo t
+
+    -- The scrutinee is itself an application
+    -- Note: we don't use the inner result type, we're going to replace it
+    , Just (SplitMatchContext innerVars (innerScrut, innerScrutAnn) _ innerBranches, innerBranchArities, reconstructInner) <- splitMatch binfo vinfo outerScrut
+    , let
+        mkNewOuterMatch newScrut =
+          reconstructOuter $ SplitMatchContext outerVars (newScrut, outerScrutAnn) (outerResTy, outerResTyAnn) outerBranches
+
+    -- See Note [Case-of-case and conapps]
+    , Just newInnerBranches <- mapBranches (\b -> if not conappsOnly || isConApp b then Just (mkNewOuterMatch b) else Nothing) innerBranches innerBranchArities
+    , let
+        newInnerMatch =
+          reconstructInner $ SplitMatchContext innerVars (innerScrut, innerScrutAnn) (outerResTy, outerResTyAnn) newInnerBranches
+
+      = newInnerMatch
```

Comment:
TBH, I don't like  all these `let`s inside the guard. It makes it hard to separate what is the "matching/condition part" and what is the "value returning/reconstruction part"